data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java

```diff
@@ -0,0 +1,92 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.beaconrestapi.handlers.v1.events;
+
+import io.javalin.http.sse.SseClient;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+
+public class EventSubscriber {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int MAX_PENDING_EVENTS = 10;
+  private final List<EventType> eventTypes;
+  private final SseClient sseClient;
+  private final Queue<QueuedEvent> queuedEvents;
+  private final Runnable closeCallback;
+  private final AtomicBoolean processingQueue;
+  final AsyncRunner asyncRunner;
+
+  public EventSubscriber(
+      final List<String> eventTypes,
+      final SseClient sseClient,
+      final Runnable closeCallback,
+      final AsyncRunner asyncRunner) {
+    this.eventTypes = getTopics(eventTypes);
+    this.sseClient = sseClient;
+    this.closeCallback = closeCallback;
+    this.queuedEvents = new ConcurrentLinkedQueue<>();
+    this.processingQueue = new AtomicBoolean(false);
+    this.asyncRunner = asyncRunner;
+    this.sseClient.onClose(closeCallback);
+  }
+
+  public void onEvent(final EventType eventType, final String message) {
+    if (eventTypes.contains(eventType)) {
+      if (queuedEvents.size() < MAX_PENDING_EVENTS) {
+        queuedEvents.add(QueuedEvent.of(eventType, message));
+        processEventQueue();
+      } else {
+        LOG.trace("Closing client connection due to exceeding the pending message limit");
+        sseClient.ctx.req.getAsyncContext().complete();
+        closeCallback.run();
+      }
+    }
+  }
+
+  static List<EventType> getTopics(List<String> topics) {
+    return topics.stream().map(EventType::valueOf).collect(Collectors.toList());
+  }
+
+  public SseClient getSseClient() {
+    return sseClient;
+  }
+
+  private void processEventQueue() {
+    if (!processingQueue.compareAndSet(false, true)) {
+      // any queue processing in progress will clear the queue, no need to run another instance
+      return;
+    }
+    asyncRunner
+        .runAsync(
+            () -> {
+              LOG.trace(
+                  "Processing queue with {} elements for event client {}",
+                  queuedEvents.size(),
+                  sseClient.hashCode());
+              while (queuedEvents.size() > 0) {
+                QueuedEvent event = queuedEvents.peek();
+                queuedEvents.remove(event);
+                sseClient.sendEvent(event.getEventType().name(), event.getMessageData());
+              }
+              processingQueue.set(false);
+            })
+        .reportExceptions();
```

Comment:
Probably safer to reset this with:
```suggestion
            })
        .alwaysRun(() -> processingQueue.set(false))
        .reportExceptions();
```

Reply:
Probably worth adding a test where `sseClient.sendEvent` throws to make sure we don't get stuck.

Reply:
It doesn't actually throw, I found out while writing this - i'll remove the .reportExceptions, and put in the alwaysRun just in case anything weird happens.
If the queue fills and we disconnect the client it'll reconnect and become a new client, so it's not the end of the world I guess.